src/stateful-set.ts

```diff
@@ -127,16 +130,44 @@ export class StatefulSet extends workload.Workload {
       metadata: props.metadata,
       spec: Lazy.any({ produce: () => this._toKube() }),
     });
-    this._service = props.service;
+    this.service = props.service ?? this.createHeadlessService();
 
-    this.apiObject.addDependency(this._service);
+    this.apiObject.addDependency(this.service);
 
     this.replicas = props.replicas ?? 1;
     this.strategy = props.strategy ?? StatefulSetUpdateStrategy.rollingUpdate(),
     this.podManagementPolicy = props.podManagementPolicy ?? PodManagementPolicy.ORDERED_READY;
     this.minReady = props.minReady ?? Duration.seconds(0);
 
-    this._service.select(this);
+    this.service.select(this);
+  }
+
+  private createHeadlessService() {
```

Comment:
⛏️ Not sure what our rule is on appending `_` and if it goes on all private methods/props

```suggestion
  private _createHeadlessService() {
```